Live: support real time measurements (alpha)

ryantxu opened the pull request:
This PR adds an initial alpha implementation for generic real-time measurement support.

This PR currently supports:
1. POST a measurement batch to the `api/live/grafana/measurements/{path}` endpoint
1. the values are broadcast to everyone listening on this channel
1. the front end will buffer these points in a circular buffer.  
1. values can be added to panels using the `-- Grafana --` datasource
1. improves the (alpha) live panel, so we can send data to channels

This is still early and experimental, however I think it is worth merging for 7.3.  The interface that should get the most attention is:
https://github.com/grafana/grafana/pull/28022/files#diff-1fbde96172d0110468fcb647d365f52729f61cf72856c91ab6113d611a946b1cR33

There is one *massive* issue with streaming as currently implemented.  We do not yet handle unsubscribing!  We *could* unsubscribe when the rxjs observable is done -- however, this is not great for lots of standard features where you navigate from one dashboard to another and expect to see the same data but displayed differently.

The easiest way to see somethign is to use the "Grafana Live" option under testdata:  This will add a point to the graph every 2s

![image](https://user-images.githubusercontent.com/705951/96173894-5dd91080-0edd-11eb-9931-f681d30b7589.png)

ryantxu committed:
improve reduce transformer

ryantxu committed:
add measurment classes

ryantxu committed:
sync with new grafana measure format

ryantxu committed:
use address for live

ryantxu committed:
use plural in URL

ryantxu committed:
set the field name

ryantxu committed:
fix build

ryantxu committed:
find changes

torkelo committed:
Yarn: Update lock file (#28014)

ivanahuckova committed:
Loki: Run instant query only in Explore (#27974)

* Run instant query only in Explore

* Replace forEach with for loop

ryantxu committed:
POST http to channel

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

ryantxu committed:
don't cast

ryantxu committed:
merge master

ozhuang committed:
Docs: Fixed row display in table (#28031)

aknuds1 committed:
Plugins: Let descendant plugins inherit their root's signature (#27970)

* plugins: Let descendant plugins inherit their root's signature

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

ahlaw committed:
Registry: Fix service shutdown mode trigger location (#28025)

Clarity-89 committed:
Add Alex Khomenko as member (#28032)

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

seanlaff commented:
Excellent! :)

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

ryantxu committed:
show history

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

ryantxu committed:
fix confirm

ryantxu committed:
fix confirm

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

ryantxu committed:
add tests

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

ryantxu committed:
fix lint

ryantxu committed:
add more errors

ryantxu committed:
merge master

ryantxu committed:
set values

ryantxu committed:
merge master

ryantxu committed:
Merge remote-tracking branch 'origin/improve-reduce' into live-measurments

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

ryantxu committed:
remove unrelated changes

ryantxu commented on packages/grafana-runtime/src/measurement/query.ts:
```diff
@@ -0,0 +1,72 @@
+/**
+ * When you know the stream will be managed measurments
+ */
+export function getLiveMeasurmentsObserver(
```
@hugohaggmark this is the function I think needs the most attention.

ryantxu commented on pkg/api/api.go:
```diff
@@ -388,6 +388,11 @@ func (hs *HTTPServer) registerRoutes() {
 
+		// The live handler will manage its own routes
+		if hs.Live != nil {
+			apiRoute.Any("/live/*", hs.LiveHTTPHandler)
```
I tried using:
```suggestion
			apiRoute.Any("/live/*", Wrap(hs.LiveHTTPHandler))
```
but the circular imports to use any response helpers was too difficult to sort out

hugohaggmark commented on packages/grafana-runtime/src/measurement/query.ts:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  rsp.key = requestId;
+  return live
+    .getChannel<LiveMeasurements>(addr)
+    .getStream()
+    .pipe(
+      map(evt => {
+        if (isLiveChannelMessageEvent(evt)) {
+          rsp.data = evt.message.getData(query);
+          if (!rsp.data.length) {
+            // ?? skip when data is empty ???
+          }
+          delete rsp.error;
+          rsp.state = LoadingState.Streaming;
+        } else if (isLiveChannelStatusEvent(evt)) {
+          if (evt.error != null) {
+            rsp.error = rsp.error;
+            rsp.state = LoadingState.Error;
+          }
+        }
+        return { ...rsp }; // send event on all status messages
+      })
+    );
```
nit: this looks fine but everytime you introduce an if statement in RxJs operator you should always ask yourself if that operator can be divided into 2 streams using `filter` and then merge the streams in the return.

torkelo replied:
why is that preferable? Seems more complex & inefficient? Can understand it if you after the filtering need to do different async stream processing

hugohaggmark replied:
I can see you point Torkel and maybe this is not the perfect use case because of efficiency and because we don't do anything combined. But usually the streams tend to grow as all other code and then I've found that keeping the logic outside the map functions is easier to read and easier to refactor. 

I was just trying to share some of the insights I found working with larger RxJs implementations, I hold no strong opinions about this.

hugohaggmark commented on public/app/plugins/datasource/grafana/datasource.ts:
```diff
@@ -6,54 +6,63 @@ import {
+      }
+    }
+    // With a single query just return the results
+    if (queries.length === 1) {
+      return queries[0];
+    }
+    if (queries.length > 1) {
+      // HELP!!!
+      return queries[0];
+    }
+    return of(); // nothing
```
Maybe something like this?
```suggestion
     return forkJoin(queries).pipe(map((results:DataFrame[]) => ... )))
```

hugohaggmark replied:
I did something similar in https://github.com/grafana/grafana/pull/28221 so maybe we should brea this out into a utility function?

hugohaggmark commented on public/app/core/components/TransformersUI/ReduceTransformerEditor.tsx:
```diff
@@ -1,42 +1,103 @@
   options,
   onChange,
 }) => {
+  const modes: Array<SelectableValue<ReduceTransformerMode>> = [
+    {
+      label: 'Series to rows',
+      value: ReduceTransformerMode.SeriesToRows,
+      description: 'Create a table with one row for each series value',
+    },
+    {
+      label: 'Reduce fields',
+      value: ReduceTransformerMode.ReduceFields,
+      description: 'Collapse each field into a single value',
+    },
+  ];
+
+  const onSelectMode = useCallback(
+    (value: SelectableValue<ReduceTransformerMode>) => {
+      const mode = value.value!;
+      onChange({
+        ...options,
+        mode,
+        includeTimeField: mode === ReduceTransformerMode.ReduceFields ? !!options.includeTimeField : false,
+      });
+    },
+    [onChange, options]
+  );
+
+  const onToggleTime = useCallback(() => {
+    onChange({
+      ...options,
+      includeTimeField: !options.includeTimeField,
+    });
+  }, [onChange, options]);
```
I don't understand how changes in this file relates to changes in live? Maybe open in a separate PR?

ryantxu replied:
dooh -- that is part of https://github.com/grafana/grafana/pull/27875

these three PRs are all required to make some good demos work ;)

hugohaggmark commented on public/app/plugins/panel/live/LiveChannelEditor.tsx:
```diff
@@ -135,7 +141,7 @@ export class LiveChannelEditor extends PureComponent<Props, State> {
               <Select
                 options={paths}
-                value={paths.find(s => s.value === path) || path || ''}
+                value={paths.find(s => s.value === path) ?? (path ? { label: path, value: path } : undefined)}
```
nit: maybe break out this value entries into functions?

ryantxu committed:
unrelated changes

hugohaggmark reviewed (commented):
This is massive work but I can't follow all the changes because I don't understand the overall picture :) so I think this PR could probably be broken up into smaller pieces or maybe it would be easier to follow if there were steps on how to test this "live" in Grafana?

Is this something that is being demoed at ObsCon because at some places it doesn't feel that clean and it doesn't build?

aknuds1 commented:
> This is massive work but I can't follow all the changes because I don't understand the overall picture :) so I think this PR could probably be broken up into smaller pieces or maybe it would be easier to follow if there were steps on how to test this "live" in Grafana?
> 
> Is this something that is being demoed at ObsCon because at some places it doesn't feel that clean and it doesn't build?

I'm also wondering if this has to go into 7.3.0-beta1 already, since it's a huge PR.

aknuds1 commented on pkg/api/live_handler.go:
```diff
@@ -0,0 +1,58 @@
+	c.JSON(404, util.DynMap{})
+}
+
+func getChannelAddress(path string) (live.ChannelAddress, bool) {
+	idx := strings.Index(path, "/api/live/")
+	if idx < 0 {
+		return live.ChannelAddress{}, false
+	}
+	return live.ParseChannelAddress(path[(idx + len("/api/live/")):]), true
+}
```
Can we just inline this function? LiveHTTPHandler is already relatively short and I don't see any benefit from the abstraction.

aknuds1 commented on pkg/models/live.go:
```diff
@@ -15,12 +15,18 @@ type ChannelHandler interface {
 
-// ChannelHandlerProvider -- this should be implemented by any core feature
-type ChannelHandlerProvider interface {
+// ChannelNamespaceHandler -- this should be implemented by any core feature
```
Documentation comments should be sentences.
```suggestion
// ChannelNamespaceHandler should be implemented by any core feature.
```

aknuds1 commented on pkg/models/live.go:
```diff
@@ -15,12 +15,18 @@ type ChannelHandler interface {
 	// This is called fast and often -- it must be synchrnozed
 	GetHandlerForPath(path string) (ChannelHandler, error)
+
+	// Called from the HTTP API
```
```suggestion
	// DoNamespaceHTTP is called from the HTTP API
```

aknuds1 commented on pkg/models/measurement.go:
```diff
@@ -0,0 +1,35 @@
+// Measurement is a single measurement value
+type Measurement struct {
+	Name   string                      `json:"name,omitempty"`
+	Time   int64                       `json:"time,omitempty"`   // units are usually ms, but depend on the channel
+	Values map[string]interface{}      `json:"values,omitempty"` // typically number or string
+	Config map[string]data.FieldConfig `json:"config,omitempty"` // optional list of field configs
+	Labels map[string]string           `json:"labels,omitempty"` // labels are applied to all values
```
The comments should go above and be on sentence form. When we enable better linting, it will catch this.

aknuds1 commented on pkg/models/measurement.go:
```diff
@@ -0,0 +1,35 @@
+
+// MeasurementBatch is a collection of measurments all sent at once
+type MeasurementBatch struct {
+	Action   MeasurementAction `json:"action,omitempty"`   // default action is append
+	Measures []Measurement     `json:"measures,omitempty"` // batch of measurments
+	Capacity int64             `json:"capacity,omitempty"` // The suggested size of the client buffer
```
The comments should go above and be on sentence form. When we enable better linting, it will catch this.

aknuds1 commented on pkg/services/live/channel.go:
```diff
@@ -1,27 +1,40 @@
-	Path      string // path within the channel handler
+// ChannelAddress is the channel id split by parts
+type ChannelAddress struct {
+	Scope     string `json:"scope,omitempty"`     // grafana, ds, or plugin
+	Namespace string `json:"namespace,omitempty"` // feature, id, or name
+	Path      string `json:"path,omitempty"`      // path within the channel handler
```
The comments should go above and be on sentence form. When we enable better linting, it will catch this.

aknuds1 commented on pkg/services/live/channel.go:
```diff
@@ -1,27 +1,40 @@
+}
+
+// IsValid checks if all parts of the address are valid
+func (id *ChannelAddress) IsValid() bool {
+	return id.Scope != "" && id.Namespace != "" && id.Path != ""
+}
```
The receiver name `id` is confusing here.
```suggestion
func (ca *ChannelAddress) IsValid() bool {
	return ca.Scope != "" && ca.Namespace != "" && ca.Path != ""
}
```

aknuds1 commented on pkg/services/live/channel.go:
```diff
@@ -1,27 +1,40 @@
+}
+
+// ToChannelID converts this to a single string
+func (id *ChannelAddress) ToChannelID() string {
```
The receiver name id is confusing here.
```suggestion
func (ca *ChannelAddress) ToChannelID() string {
```

aknuds1 commented on pkg/services/live/channel.go:
```diff
@@ -1,27 +1,40 @@
+
+// ToChannelID converts this to a single string
+func (id *ChannelAddress) ToChannelID() string {
+	return id.Scope + "/" + id.Namespace + "/" + id.Path
```
```suggestion
	return ca.Scope + "/" + ca.Namespace + "/" + ca.Path
```

aknuds1 commented on pkg/services/live/features/broadcast.go:
```diff
@@ -1,32 +1,86 @@
+	return b, nil // for now all channels share config
+}
+
+// DoNamespaceHTTP called from the http api
```
```suggestion
// DoNamespaceHTTP is called from the HTTP API.
```

aknuds1 commented on pkg/services/live/features/broadcast.go:
```diff
@@ -1,32 +1,86 @@
+// DoNamespaceHTTP called from the http api
+func (b *BroadcastRunner) DoNamespaceHTTP(c *models.ReqContext) {
+	c.JSON(400, util.DynMap{
+		"Unsupportedd": "BroadcastRunner",
```
```suggestion
		"Unsupported": "BroadcastRunner",
```

aknuds1 commented on pkg/services/live/features/broadcast.go:
```diff
@@ -1,32 +1,86 @@
+
+// DoChannelHTTP accepts POST from anywhere
+func (b *BroadcastRunner) DoChannelHTTP(c *models.ReqContext, channel string) {
+	if c.Req.Method == "POST" {
```
It's better/more idiomatic to invert the `if` here and fail and return if the method isn't "POST". It will make the code more readable due to reduced nesting.

aknuds1 commented on pkg/services/live/features/dashboard.go:
```diff
@@ -17,21 +18,21 @@ type dashboardEvent struct {
 	return g, nil // all dashboards share the same handler
 }
 
+// DoNamespaceHTTP called from the http api
```
```suggestion
// DoNamespaceHTTP is called from the HTTP API.
```

aknuds1 commented on pkg/services/live/features/dashboard.go:
```diff
@@ -52,13 +53,20 @@ func (g *DashboardHandler) OnPublish(c *centrifuge.Client, e centrifuge.PublishE
 	return e.Data, nil
 }
 
+// DoChannelHTTP called from the http api
```
```suggestion
// DoChannelHTTP is called from the HTTP API.
```

aknuds1 commented on pkg/services/live/features/dashboard.go:
```diff
@@ -17,21 +18,21 @@ type dashboardEvent struct {
+// DoNamespaceHTTP called from the http api
+func (g *DashboardHandler) DoNamespaceHTTP(c *models.ReqContext) {
+	c.JSON(400, util.DynMap{
+		"Unsupportedd": "DashboardHandler",
```
```suggestion
		"Unsupported": "DashboardHandler",
```

aknuds1 commented on pkg/services/live/features/dashboard.go:
```diff
@@ -52,13 +53,20 @@ func (g *DashboardHandler) OnPublish(c *centrifuge.Client, e centrifuge.PublishE
+// DoChannelHTTP called from the http api
+func (g *DashboardHandler) DoChannelHTTP(c *models.ReqContext, channel string) {
+	c.JSON(400, util.DynMap{
+		"Unsupportedd": channel,
```
```suggestion
		"Unsupported": channel,
```

aknuds1 commented on pkg/services/live/features/measurements.go:
```diff
@@ -0,0 +1,92 @@
+	Publisher models.ChannelPublisher
+}
+
+// GetHandlerForPath called on init
```
```suggestion
// GetHandlerForPath is called on init.
```

aknuds1 commented on pkg/services/live/features/measurements.go:
```diff
@@ -0,0 +1,92 @@
+	return m, nil // for now all channels share config
+}
+
+// DoNamespaceHTTP called from the http api
```
```suggestion
// DoNamespaceHTTP is called from the HTTP API.
```

aknuds1 commented on pkg/services/live/features/measurements.go:
```diff
@@ -0,0 +1,92 @@
+// DoNamespaceHTTP called from the http api
+func (m *MeasurementsRunner) DoNamespaceHTTP(c *models.ReqContext) {
+	c.JSON(400, util.DynMap{
+		"Unsupportedd": "MeasurementsRunner",
```
```suggestion
		"Unsupported": "MeasurementsRunner",
```

aknuds1 commented on pkg/services/live/features/measurements.go:
```diff
@@ -0,0 +1,92 @@
+	})
+}
+
+// GetChannelOptions called fast and often
```
```suggestion
// GetChannelOptions gets channel options.
// It gets called fast and often.
```

aknuds1 commented on pkg/services/live/features/measurements.go:
```diff
@@ -0,0 +1,92 @@
+	return nil
+}
+
+// OnPublish called when an event is received from the websocket
```
```suggestion
// OnPublish is called when an event is received from the websocket.
```

aknuds1 commented on pkg/services/live/features/measurements.go:
```diff
@@ -0,0 +1,92 @@
+	return e.Data, nil
+}
+
+// DoChannelHTTP called from the http api
```
```suggestion
// DoChannelHTTP is called from the HTTP API.
```

aknuds1 commented on pkg/services/live/features/measurements.go:
```diff
@@ -0,0 +1,92 @@
+
+// DoChannelHTTP called from the http api
+func (m *MeasurementsRunner) DoChannelHTTP(c *models.ReqContext, channel string) {
+	if c.Req.Method == "POST" {
```
See my previous comment about returning early on error.

aknuds1 commented on pkg/services/live/features/measurements.go:
```diff
@@ -0,0 +1,92 @@
+		err = json.Unmarshal(body, &msg)
+		if err != nil {
+			c.JSON(500, util.DynMap{
+				"message": "body must be measurment batch",
```
```suggestion
				"message": "body must be measurement batch",
```

aknuds1 commented on pkg/services/live/features/measurements.go:
```diff
@@ -0,0 +1,92 @@
+	}
+
+	c.JSON(400, util.DynMap{
+		"unsuppoted?": channel,
```
```suggestion
		"unsupported?": channel,
```

aknuds1 commented on pkg/services/live/features/testdata.go:
```diff
@@ -18,18 +18,19 @@ type testdataRunner struct {
+// DoNamespaceHTTP called from the http api
+func (g *TestdataSupplier) DoNamespaceHTTP(c *models.ReqContext) {
+	c.JSON(400, util.DynMap{
+		"Unsupportedd": "TestdataSupplier",
```
```suggestion
		"Unsupported": "TestdataSupplier",
```

aknuds1 commented on pkg/services/live/features/testdata.go:
```diff
@@ -82,11 +84,11 @@ func (g *testdataRunner) OnPublish(c *centrifuge.Client, e centrifuge.PublishEve
+// DoChannelHTTP called from the http api
+func (g *testdataRunner) DoChannelHTTP(c *models.ReqContext, channel string) {
+	c.JSON(400, util.DynMap{
+		"Unsupportedd": channel,
```
```suggestion
		"Unsupported": channel,
```

aknuds1 commented on pkg/services/live/features/testdata.go:
```diff
@@ -105,20 +114,20 @@ func (g *testdataRunner) runRandomCSV() {
 		err = g.publisher(g.channel, bytes)
 		if err != nil {
-			logger.Warn("write", "channel", g.channel, "line", line)
+			logger.Warn("write", "channel", g.channel, "line", measure)
```
```suggestion
			logger.Warn("write", "channel", g.channel, "measure", measure)
```

aknuds1 commented on pkg/services/live/live.go:
```diff
@@ -245,39 +253,50 @@ func (g *GrafanaLive) GetChannelHandler(channel string) (models.ChannelHandler,
+			return p, nil
 		}
-		return nil, fmt.Errorf("Unknown feature: %s", id.Namespace)
+		return nil, fmt.Errorf("Unknown feature: %s", name)
```
I think quoting helps debugging, in case of spaces in name.
```suggestion
		return nil, fmt.Errorf("Unknown feature: %q", name)
```

aknuds1 commented on pkg/services/live/live.go:
```diff
@@ -245,39 +253,50 @@ func (g *GrafanaLive) GetChannelHandler(channel string) (models.ChannelHandler,
-	if id.Scope == "ds" {
-		return nil, fmt.Errorf("todo... look up datasource: %s", id.Namespace)
+	if scope == "ds" {
+		return nil, fmt.Errorf("todo... look up datasource: %s", name)
```
```suggestion
		return nil, fmt.Errorf("todo... look up datasource: %q", name)
```

aknuds1 commented on pkg/services/live/live.go:
```diff
@@ -245,39 +253,50 @@ func (g *GrafanaLive) GetChannelHandler(channel string) (models.ChannelHandler,
+			return h, nil
 		}
-		return nil, fmt.Errorf("unknown plugin: %s", id.Namespace)
+		return nil, fmt.Errorf("unknown plugin: %s", name)
```
```suggestion
		return nil, fmt.Errorf("unknown plugin: %q", name)
```

aknuds1 commented on pkg/services/live/live.go:
```diff
@@ -245,39 +253,50 @@ func (g *GrafanaLive) GetChannelHandler(channel string) (models.ChannelHandler,
+		return nil, fmt.Errorf("unknown plugin: %s", name)
+	}
+
+	return nil, fmt.Errorf("invalid scope: %s", scope)
```
```suggestion
	return nil, fmt.Errorf("invalid scope: %q", scope)
```

aknuds1 commented on pkg/services/live/pluginHandler.go:
```diff
@@ -16,6 +17,13 @@ func (g *PluginHandler) GetHandlerForPath(path string) (models.ChannelHandler, e
 	return g, nil // all dashboards share the same handler
 }
 
+// DoNamespaceHTTP called from the http api
```
```suggestion
// DoNamespaceHTTP is called from the HTTP API.
```

aknuds1 commented on pkg/services/live/pluginHandler.go:
```diff
@@ -16,6 +17,13 @@ func (g *PluginHandler) GetHandlerForPath(path string) (models.ChannelHandler, e
+// DoNamespaceHTTP called from the http api
+func (g *PluginHandler) DoNamespaceHTTP(c *models.ReqContext) {
+	c.JSON(400, util.DynMap{
+		"Unsupportedd": "TestdataSupplier",
```
```suggestion
		"Unsupported": "TestdataSupplier",
```

aknuds1 commented on pkg/services/live/pluginHandler.go:
```diff
@@ -30,3 +38,10 @@ func (g *PluginHandler) OnSubscribe(c *centrifuge.Client, e centrifuge.Subscribe
 	return e.Data, nil // broadcast any event
 }
+
+// DoChannelHTTP called when an event is received from the websocket
+func (g *PluginHandler) DoChannelHTTP(c *models.ReqContext, channel string) {
+	c.JSON(400, util.DynMap{
+		"Unsupportedd": channel,
+	})
```
```suggestion
// DoChannelHTTP is called when an event is received from the websocket.
func (g *PluginHandler) DoChannelHTTP(c *models.ReqContext, channel string) {
	c.JSON(400, util.DynMap{
		"Unsupported": channel,
	})
```

aknuds1 reviewed (commented):
I have to admit I don't understand yet what this does, and I would like to be able to, but see my comments/suggestions for now.

marefr commented:
Does this aiming to close #4355?

torkelo commented:
yes, it's very much that

ryantxu committed:
Update pkg/models/live.go

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

ryantxu committed:
Update pkg/models/live.go

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

ryantxu committed:
Update pkg/services/live/live.go

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

ryantxu committed:
Update pkg/services/live/pluginHandler.go

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

ryantxu committed:
Update pkg/services/live/pluginHandler.go

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

ryantxu committed:
Update pkg/services/live/pluginHandler.go

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

ryantxu committed:
use measurments for testdata endpoints

ryantxu committed:
add live to testdata

ryantxu committed:
add live to testdata

ryantxu committed:
Update pkg/services/live/channel.go

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

ryantxu committed:
Apply suggestions from code review

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

ryantxu committed:
update comment formats

ryantxu committed:
uprevert testdata

nonamef committed:
CloudWatch: Add EC2CapacityReservations Namespace (#28309)

aknuds1 committed:
API: Fix short URLs (#28300)

* API: Fix short URLs

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Chore: Add cloud-middleware as code owners (#28310)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

papagian committed:
SQLStore: Run tests as integration tests (#28265)

* sqlstore: Run tests as integration tests

* Truncate database instead of re-creating it on each test

* Fix test description

See https://github.com/grafana/grafana/pull/12129

* Fix lint issues

* Fix postgres dialect after review suggestion

* Rename and document functions after review suggestion

* Add periods

* Fix auto-increment value for mysql dialect

Co-authored-by: Emil Tullstedt <emil.tullstedt@grafana.com>

aknuds1 committed:
Drone: Fix grafana-mixin linting (#28308)

* Drone: Fix Starlark script

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

* grafana-mixin: Move build logic to scripts

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

* Drone: Use mixin scripts

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

* CI build image: Install jsonnetfmt and mixtool

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

* Makefile: Print commands

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

bergquist committed:
should only ignore the file in the grafana mixin root folder (#28306)

Signed-off-by: bergquist <carl.bergquist@gmail.com>

jackw committed:
fix: for graph size not taking up full height or width

dprokop committed:
Graph NG: fix toggling queries and extract Graph component from graph3 panel (#28290)

* Fix issue when data and config is not in sync

* Extract GraphNG component from graph panel and add some tests coverage

* Update packages/grafana-ui/src/components/uPlot/hooks.test.ts

* Update packages/grafana-ui/src/components/uPlot/hooks.test.ts

* Update packages/grafana-ui/src/components/uPlot/hooks.test.ts

* Fix grid color and annotations refresh

aknuds1 committed:
Drone: Use ${DRONE_TAG} in release pipelines, since it should work (#28299)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

elliotpryde committed:
Explore: respect min_refresh_interval (#27988)

* Explore: respect min_refresh_interval

Fixes #27494

* fixup! Explore: respect min_refresh_interval

* fixup! Explore: respect min_refresh_interval

* UI: export defaultIntervals from refresh picker

* fixup! Explore: respect min_refresh_interval

Co-authored-by: Zoltán Bedi <zoltan.bedi@gmail.com>

aocenas committed:
Loki: Base maxDataPoints limits on query type (#28298)

* Base maxLines and maxDataPoints based on query type

* Allow overriding the limit to higher value

dependabot[bot] committed:
Bump tree-kill from 1.2.1 to 1.2.2 (#27405)

Bumps [tree-kill](https://github.com/pkrumins/node-tree-kill) from 1.2.1 to 1.2.2.
- [Release notes](https://github.com/pkrumins/node-tree-kill/releases)
- [Commits](https://github.com/pkrumins/node-tree-kill/compare/v1.2.1...v1.2.2)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump handlebars from 4.4.3 to 4.7.6 (#27416)

Bumps [handlebars](https://github.com/wycats/handlebars.js) from 4.4.3 to 4.7.6.
- [Release notes](https://github.com/wycats/handlebars.js/releases)
- [Changelog](https://github.com/handlebars-lang/handlebars.js/blob/master/release-notes.md)
- [Commits](https://github.com/wycats/handlebars.js/compare/v4.4.3...v4.7.6)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Build(deps): Bump http-proxy from 1.18.0 to 1.18.1 (#27507)

Bumps [http-proxy](https://github.com/http-party/node-http-proxy) from 1.18.0 to 1.18.1.
- [Release notes](https://github.com/http-party/node-http-proxy/releases)
- [Changelog](https://github.com/http-party/node-http-proxy/blob/master/CHANGELOG.md)
- [Commits](https://github.com/http-party/node-http-proxy/compare/1.18.0...1.18.1)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

torkelo committed:
Automation: Add backport github action (#28318)

hugohaggmark committed:
BackendSrv: Fixes queue countdown when unsubscribe is before response (#28323)

dprokop committed:
GraphNG: Use AxisSide enum (#28320)

torkelo committed:
IssueTriage: Needs more info automation and messages (#28137)

* IssueTriage: Needs more info automation and messages

* Updated

* Updated

* Updated wording

alexanderzobnin committed:
SAML: IdP-initiated SSO docs (#28280)

* SAML: IdP-initiated SSO docs

* Update docs/sources/enterprise/saml.md

Co-authored-by: Emil Tullstedt <emil.tullstedt@grafana.com>

* Apply suggestions from code review

Co-authored-by: Emil Tullstedt <emil.tullstedt@grafana.com>

Co-authored-by: Emil Tullstedt <emil.tullstedt@grafana.com>

aocenas committed:
Loki: Run instant query only when doing metric query (#28325)

* Run instant query only when doing metric query

* Update public/app/plugins/datasource/loki/datasource.ts

Co-authored-by: Ivana Huckova <30407135+ivanahuckova@users.noreply.github.com>

Co-authored-by: Ivana Huckova <30407135+ivanahuckova@users.noreply.github.com>

torkelo committed:
Automation: Tweaks to more info message (#28332)

kylebrandt committed:
AlertingNG: remove warn/crit from eval prototype (#28334)

and misc cleanup

briangann committed:
area/grafana/toolkit: update e2e docker image (#28335)

* add xvfb to image

* comment out toolkit inclusion

* add latest tag

* update packages for cypress

* cleanup script

J-F-Far committed:
Update auth-proxy.md (#28339)

Fix a minor grammar mistake: 'handling' to 'handle'.

acoder77 committed:
Git: Create .gitattributes for windows line endings (#28340)

With this set, Windows users will have text files converted from Windows style line endings (\r\n) to Unix style line endings (\n) when they’re added to the repository.
https://www.edwardthomson.com/blog/git_for_windows_line_endings.html

peterholmberg committed:
Docs: Add docs for valuepicker (#28327)

krzysdabro committed:
Templating: Replace all '$tag' in tag values query (#28343)

papagian committed:
Docs: Add missing records from grafana-ui 7.2.1 CHANGELOG (#28302)

maknik committed:
Dashboard links: Places drop down list so it's always visible (#28330)

* calculating whether to place the list on the right or left edge of the parent

* change naming and add import of createRef

torkelo committed:
Automation: Update backport github action trigger (#28352)

It seems like GitHub has solved the problem of running github actions on PRs from forks with access to secrets. 

https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/#improvements-for-public-repository-forks

If I change the event that triggers it to pull_request_target the action is run in the context of the base instead of the merged PR branch

torkelo committed:
ColorSchemes: Adds more color schemes and text colors that depend on the background (#28305)

* Adding more color modes and text colors that depend on the background color

* Updates

* Updated

* Another big value fix

* Fixing unit tests

* Updated

* Updated test

* Update

* Updated

* Updated

* Updated

* Updated

* Added new demo dashboard

* Updated

* updated

* Updated

* Updateed

* added beta notice

* Fixed e2e test

wbrowne commented on public/app/plugins/datasource/grafana/datasource.ts:
```diff
@@ -105,3 +114,32 @@ export class GrafanaDatasource extends DataSourceApi<GrafanaQuery> {
+function getRandomWalk(request: DataQueryRequest): Observable<DataQueryResponse> {
+  const { intervalMs, maxDataPoints, range, requestId } = request;
+
+  // Yes, this implementaiton ignores multiple targets!  But that matches exisitng behavior
```
```suggestion
  // Yes, this implementation ignores multiple targets!  But that matches existing behavior
```

wbrowne commented on public/app/plugins/panel/live/module.tsx:
```diff
@@ -1,14 +1,37 @@
+    .addRadio({
+      path: 'message',
+      name: 'Show Message',
+      description: 'Display the last message recieved on this channel',
```
```suggestion
      description: 'Display the last message received on this channel',
```

wbrowne commented on public/app/plugins/panel/live/module.tsx:
```diff
@@ -1,14 +1,37 @@
+    .addBooleanSwitch({
+      path: 'publish',
+      name: 'Show Publish',
+      description: 'Display the a form to publish values',
```
Not sure which is meant here - `Display the form to publish values` / `Display a form to publish values`?

ryantxu replied:
dooh -- either, but not both!

wbrowne reviewed (commented):
Have tested a bit and it seems to work as expected. Will continue testing for a bit more.

Found some spelling/grammar mistakes in the meantime

aknuds1 commented on pkg/models/live.go:
```diff
@@ -15,12 +15,18 @@ type ChannelHandler interface {
 	// Called when something writes into the channel.  The returned value will be broadcast if len() > 0
 	OnPublish(c *centrifuge.Client, e centrifuge.PublishEvent) ([]byte, error)
+
+	// Called from the HTTP API
```
```suggestion
	// DoChannelHTTP is called from the HTTP API
```

aknuds1 commented on pkg/models/measurement.go:
```diff
@@ -0,0 +1,50 @@
+import "github.com/grafana/grafana-plugin-sdk-go/data"
+
+// NOTE:
+// this likely shoud go in the Plugin SDK since it will be useful from plugins
```
```suggestion
// this likely should go in the Plugin SDK since it will be useful from plugins
```

aknuds1 commented on pkg/models/measurement.go:
```diff
@@ -0,0 +1,50 @@
+
+// Measurement is a single measurement value
+type Measurement struct {
+	// the measurement name
```
```suggestion
	// Name of the measurement.
```

aknuds1 commented on pkg/models/measurement.go:
```diff
@@ -0,0 +1,50 @@
+	// the measurement name
+	Name string `json:"name,omitempty"`
+
+	// units are usually ms, but depend on the channel
```
```suggestion
	// Time is the measurement time. Units are usually ms, but depends on the channel
```

aknuds1 commented on pkg/models/measurement.go:
```diff
@@ -0,0 +1,50 @@
+	// units are usually ms, but depend on the channel
+	Time int64 `json:"time,omitempty"`
+
+	// typically number or string
```
```suggestion
	// Values is the measurement's values. The value type is typically number or string.
```

aknuds1 commented on pkg/models/measurement.go:
```diff
@@ -0,0 +1,50 @@
+	// typically number or string
+	Values map[string]interface{} `json:"values,omitempty"`
+
+	// optional list of field configs
```
```suggestion
	// Config is an optional list of field configs.
```

aknuds1 commented on pkg/models/measurement.go:
```diff
@@ -0,0 +1,50 @@
+	// optional list of field configs
+	Config map[string]data.FieldConfig `json:"config,omitempty"`
+
+	// labels are applied to all values
```
What is the key and value in this map?
```suggestion
	// Labels are applied to all values.
```

aknuds1 commented on pkg/models/measurement.go:
```diff
@@ -0,0 +1,50 @@
+
+// MeasurementBatch is a collection of measurments all sent at once
+type MeasurementBatch struct {
+	// default action is append
```
```suggestion
	// Action is the action in question, the default is append.
```

aknuds1 commented on pkg/models/measurement.go:
```diff
@@ -0,0 +1,50 @@
+	// default action is append
+	Action MeasurementAction `json:"action,omitempty"`
+
+	// batch of measurments
```
```suggestion
	// Measures is the array of measurements.
```

aknuds1 commented on pkg/models/measurement.go:
```diff
@@ -0,0 +1,50 @@
+	Action MeasurementAction `json:"action,omitempty"`
+
+	// batch of measurments
+	Measures []Measurement `json:"measures,omitempty"`
```
Should this field be renamed to measurements, to be consistent?? The inconsistent terminology is confusing.

aknuds1 commented on pkg/models/measurement.go:
```diff
@@ -0,0 +1,50 @@
+	// batch of measurments
+	Measures []Measurement `json:"measures,omitempty"`
+
+	// The suggested size of the client buffer
```
```suggestion
	// Capacity is the suggested size of the client buffer
```

aknuds1 reviewed (changes requested):
See suggestions.

aknuds1 commented on pkg/api/live_handler.go:
```diff
@@ -0,0 +1,58 @@
+	"github.com/grafana/grafana/pkg/util"
+)
+
+// LiveHTTPHandler will process any requets to /api/live/*
```
```suggestion
// LiveHTTPHandler will process any requests to /api/live/*.
```

aknuds1 commented:
@ryantxu How can one test this PR so the `LiveHTTPHandler` method gets hit? I tested as you describe in the PR description, i.e. via gdev-testdata/Grafana Live, but that doesn't send HTTP requests.

aknuds1 commented on pkg/api/live_handler.go:
```diff
@@ -0,0 +1,58 @@
+)
+
+// LiveHTTPHandler will process any requets to /api/live/*
+func (hs *HTTPServer) LiveHTTPHandler(c *models.ReqContext) {
```
I don't see how this handler gets exercised from the frontend, I'd need to know how it's supposed to be used before I can approve.

aknuds1 commented on pkg/services/live/channel.go:
```diff
@@ -1,27 +1,40 @@
 //   ${scope} / ${namespace} / ${path}
-func ParseChannelIdentifier(id string) (ChannelIdentifier, error) {
+func ParseChannelAddress(id string) ChannelAddress {
+	identifier := ChannelAddress{}
```
```suggestion
	addr := ChannelAddress{}
```

aknuds1 commented on pkg/services/live/channel.go:
```diff
@@ -1,27 +1,40 @@
-			Namespace: parts[1],
-			Path:      parts[2],
-		}, nil
+	length := len(parts)
+	if length > 0 {
+		identifier.Scope = parts[0]
```
```suggestion
	length := len(parts)
	if length > 0 {
		addr.Scope = parts[0]
```

aknuds1 commented on pkg/services/live/channel.go:
```diff
@@ -1,27 +1,40 @@
+		identifier.Scope = parts[0]
 	}
-	return ChannelIdentifier{}, fmt.Errorf("Invalid channel id: %s", id)
+	if length > 1 {
+		identifier.Namespace = parts[1]
+	}
+	if length > 2 {
+		identifier.Path = parts[2]
+	}
+	return identifier
```
```suggestion
	if length > 1 {
		addr.Namespace = parts[1]
	}
	if length > 2 {
		addr.Path = parts[2]
	}
	return addr
```

aknuds1 commented on pkg/services/live/live.go:
```diff
@@ -232,9 +240,9 @@ func (g *GrafanaLive) GetChannelHandler(channel string) (models.ChannelHandler,
-		return nil, err
+	id := ParseChannelAddress(channel)
+	if !id.IsValid() {
+		return nil, fmt.Errorf("invalid channel")
```
Much better for debugging if the invalid channel is included:
```suggestion
		return nil, fmt.Errorf("invalid channel: %q", channel)
```

aknuds1 commented on pkg/services/live/live.go:
```diff
@@ -245,39 +253,50 @@ func (g *GrafanaLive) GetChannelHandler(channel string) (models.ChannelHandler,
+			return p, nil
 		}
-		return nil, fmt.Errorf("Unknown feature: %s", id.Namespace)
+		return nil, fmt.Errorf("Unknown feature: %q", name)
```
```suggestion
		return nil, fmt.Errorf("unknown feature: %q", name)
```

aknuds1 commented on packages/grafana-runtime/src/measurement/types.ts:
```diff
@@ -0,0 +1,70 @@
+}
+
+/**
+ * Channels that recieve Measurements can collect them into frames
```
```suggestion
 * Channels that receive Measurements can collect them into frames
```

aknuds1 commented on public/app/features/live/features.ts:
```diff
@@ -39,7 +47,13 @@ export function registerLiveFeatures() {
       getSupportedPaths: () => [broadcastConfig],
     },
-    description: 'Broadcast will send/recieve any events on a channel',
+    description: 'Broadcast will send/recieve any JSON object in a channel',
```
```suggestion
    description: 'Broadcast will send/receive any JSON object in a channel',
```

aknuds1 commented on public/app/features/live/features.ts:
```diff
@@ -39,7 +47,13 @@ export function registerLiveFeatures() {
+
+  grafanaLiveCoreFeatures.register({
+    name: 'measurements',
+    support: new LiveMeasurmentsSupport(),
```
```suggestion
    support: new LiveMeasurementsSupport(),
```

aknuds1 commented on public/app/plugins/datasource/grafana/components/QueryEditor.tsx:
```diff
@@ -23,18 +31,137 @@ export class QueryEditor extends PureComponent<Props> {
+      measurements = {};
+    }
+    const names: Array<SelectableValue<string>> = [
+      { value: '', label: 'All measurements', description: 'show every measurment streamed to this channel' },
```
```suggestion
      { value: '', label: 'All measurements', description: 'show every measurement streamed to this channel' },
```

aknuds1 commented on public/app/plugins/datasource/grafana/components/QueryEditor.tsx:
```diff
@@ -23,18 +31,137 @@ export class QueryEditor extends PureComponent<Props> {
+              onChange={this.onChannelChange}
+              allowCustomValue={true}
+              backspaceRemovesValue={true}
+              placeholder="Select measurments channel"
```
```suggestion
              placeholder="Select measurements channel"
```

aknuds1 commented on public/app/plugins/datasource/grafana/components/QueryEditor.tsx:
```diff
@@ -23,18 +31,137 @@ export class QueryEditor extends PureComponent<Props> {
+              <Select
+                options={names}
+                value={names.find(v => v.value === measurements?.name) || names[0]}
+                onChange={this.onMeasurmentNameChanged}
```
```suggestion
                onChange={this.onMeasurementNameChanged}
```

aknuds1 commented on public/app/plugins/datasource/grafana/components/QueryEditor.tsx:
```diff
@@ -23,18 +31,137 @@ export class QueryEditor extends PureComponent<Props> {
+          </div>
+        )}
+
+        <FeatureInfoBox title="Grafana Live - Measurments" featureState={FeatureState.alpha}>
```
```suggestion
        <FeatureInfoBox title="Grafana Live - Measurements" featureState={FeatureState.alpha}>
```

aknuds1 commented on public/app/plugins/datasource/grafana/components/QueryEditor.tsx:
```diff
@@ -23,18 +31,137 @@ export class QueryEditor extends PureComponent<Props> {
+
+        <FeatureInfoBox title="Grafana Live - Measurments" featureState={FeatureState.alpha}>
+          <p>
+            This supports real-time event streams in grafana core. This feature is under heavy development. Expect the
```
```suggestion
            This supports real-time event streams Ggrafana core. This feature is under heavy development. Expect the
```

aknuds1 commented on public/app/plugins/datasource/grafana/datasource.ts:
```diff
@@ -6,54 +6,63 @@ import {
-import { GrafanaQuery, GrafanaAnnotationQuery, GrafanaAnnotationType } from './types';
-import { getBackendSrv, getTemplateSrv, TemplateSrv, toDataQueryResponse } from '@grafana/runtime';
+import { GrafanaQuery, GrafanaAnnotationQuery, GrafanaAnnotationType, GrafanaQueryType } from './types';
+import { getBackendSrv, getTemplateSrv, toDataQueryResponse, getLiveMeasurmentsObserver } from '@grafana/runtime';
```
```suggestion
import { getBackendSrv, getTemplateSrv, toDataQueryResponse, getLiveMeasurementsObserver } from '@grafana/runtime';
```

aknuds1 commented on public/app/plugins/datasource/grafana/datasource.ts:
```diff
@@ -6,54 +6,63 @@ import {
+        const { channel, measurements } = target;
+        if (channel) {
+          queries.push(
+            getLiveMeasurmentsObserver(
```
```suggestion
            getLiveMeasurementsObserver(
```

aknuds1 commented on public/app/plugins/datasource/testdata/datasource.ts:
```diff
@@ -17,9 +17,16 @@ import {
+  toDataQueryError,
+  getTemplateSrv,
+  TemplateSrv,
+  getLiveMeasurmentsObserver,
```
```suggestion
  getLiveMeasurementsObserver,
```

aknuds1 commented on packages/grafana-runtime/src/measurement/collector.test.ts:
```diff
@@ -0,0 +1,250 @@
+import { MeasurementCollector } from './collector';
+import { MeasurmentAction } from './types';
```
```suggestion
import { MeasurementAction } from './types';
```

aknuds1 commented on packages/grafana-runtime/src/measurement/collector.test.ts:
```diff
@@ -0,0 +1,250 @@
+    `);
+
+    collector.addBatch({
+      action: MeasurmentAction.Replace,
```
```suggestion
      action: MeasurementAction.Replace,
```

aknuds1 commented on packages/grafana-runtime/src/measurement/collector.test.ts:
```diff
@@ -0,0 +1,250 @@
+    `);
+
+    collector.addBatch({
+      action: MeasurmentAction.Clear,
```
```suggestion
      action: MeasurementAction.Clear,
```

aknuds1 commented on packages/grafana-runtime/src/measurement/collector.ts:
```diff
@@ -0,0 +1,203 @@
+} from '@grafana/data';
+import { Measurement, MeasurementBatch, LiveMeasurements, MeasurementsQuery, MeasurmentAction } from './types';
+
+interface MeasurmentCacheConfig {
```
I noticed the same.

aknuds1 commented on packages/grafana-runtime/src/measurement/collector.ts:
```diff
@@ -0,0 +1,203 @@
+}
+
+/** This is a cache scoped to a the measurement name */
+export class MeasurmentCache {
```
```suggestion
export class MeasurementCache {
```

aknuds1 commented on packages/grafana-runtime/src/measurement/collector.ts:
```diff
@@ -0,0 +1,203 @@
+export class MeasurmentCache {
+  readonly frames: Record<string, CircularDataFrame> = {}; // key is the labels
+
+  constructor(public name: string, private config: MeasurmentCacheConfig) {
```
```suggestion
  constructor(public name: string, private config: MeasurementCacheConfig) {
```

aknuds1 reviewed (changes requested):
Please rewrite measurment as measurement.

aknuds1 committed:
Apply suggestions from code review

Co-authored-by: Will Browne <wbrowne@users.noreply.github.com>
Co-authored-by: Ryan McKinley <ryantxu@gmail.com>
Co-authored-by: Hugo Häggmark <hugo.haggmark@grafana.com>

aknuds1 committed:
Apply suggestions from code review

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into live-measurments

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=28022) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=28022) before we can accept your contribution.<br/>**23** out of **24** committers have signed the CLA.<br/><br/>:white_check_mark: ryantxu<br/>:white_check_mark: aknuds1<br/>:white_check_mark: torkelo<br/>:white_check_mark: ozhuang<br/>:white_check_mark: ivanahuckova<br/>:white_check_mark: ahlaw<br/>:white_check_mark: papagian<br/>:white_check_mark: nonamef<br/>:white_check_mark: bergquist<br/>:white_check_mark: Clarity-89<br/>:white_check_mark: aocenas<br/>:white_check_mark: jackw<br/>:white_check_mark: elliotpryde<br/>:white_check_mark: dprokop<br/>:white_check_mark: briangann<br/>:white_check_mark: hugohaggmark<br/>:white_check_mark: kylebrandt<br/>:white_check_mark: J-F-Far<br/>:white_check_mark: alexanderzobnin<br/>:white_check_mark: maknik<br/>:white_check_mark: krzysdabro<br/>:white_check_mark: acoder77<br/>:white_check_mark: peterholmberg<br/>:x: dependabot[bot]<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=28022) it.</sub>

aknuds1 committed:
Fix typos

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 reviewed (changes requested):
I've committed my previous suggestions, so will try reviewing again tomorrow, now that I know more about how it works.

ryantxu committed:
revert pseduo code

ryantxu committed:
apply feedback

ryantxu committed:
remove HTTP for now

ryantxu committed:
fix backend test

ryantxu committed:
change to datasource

ryantxu committed:
clear input for streams

ryantxu committed:
fix docs?

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

ryantxu committed:
consistent measure vs measurements

ryantxu committed:
better jsdocs

ryantxu committed:
fix a few jsdoc errors

ryantxu commented on packages/grafana-data/src/transformations/transformers/concat.ts:
```diff
@@ -73,7 +73,7 @@ export function concatenateFields(data: DataFrame[], opts: ConcatenateTransforme
           copy.labels = { ...f.labels };
           copy.labels[frameNameLabel] = frame.name;
-        } else if (!copy.name || copy.name === 'Value') {
+        } else if (!copy.name || copy.name === TIME_SERIES_VALUE_FIELD_NAME) {
```
kinda unrelated... but it took me a while to track down why stuff was behaving weird with "value" rather than "Value"

ryantxu committed:
fix comment style

hugohaggmark commented on public/app/plugins/panel/live/LiveChannelEditor.tsx:
```diff
@@ -141,7 +141,7 @@ export class LiveChannelEditor extends PureComponent<Props, State> {
               <Select
                 options={paths}
-                value={paths.find(s => s.value === path) ?? (path ? { label: path, value: path } : undefined)}
+                value={findPathOption(paths, path)}
```
Nice!

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into live-measurments

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into live-measurments

aknuds1 committed:
Remove commented out code

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Clean up code

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Clean up code

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Clean up code

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Clean up code

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Clean up code

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

hugohaggmark commented:
> There is one _massive_ issue with streaming as currently implemented. We do not yet handle unsubscribing! We _could_ unsubscribe when the rxjs observable is done -- however, this is not great for lots of standard features where you navigate from one dashboard to another and expect to see the same data but displayed differently.

I'm not sure if this is the _massive_ issue you're mentioning Ryan but it seems like every TestData panel I add that uses Grafana Live starts pushing websocket messages and it never stops.
![live-feature](https://user-images.githubusercontent.com/562238/96587724-fb1fb480-12e2-11eb-8651-00722e6f0cd3.gif)

Changing it to another scenario will not stop it either
![live-feature2](https://user-images.githubusercontent.com/562238/96588860-6cac3280-12e4-11eb-9f0f-b7a7c21cb020.gif)

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into live-measurments

ryantxu commented:
> I'm not sure if this is the massive issue you're mentioning Ryan 

Yes -- I would like to solve this in a follow on PR.  The logic will be a little tricky and it would get lost in this one.   On the frontend, we do not want to stop listening immediately when unsubscribed because we need to keep state on refresh and when navigating from page to page.  My plan is to add a watchdog check that will cleanup any open connections that are not listened to

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into live-measurments

aknuds1 committed:
Clean up code

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 commented on pkg/models/live.go:
```diff
@@ -17,9 +17,10 @@ type ChannelHandler interface {
-// ChannelHandlerProvider -- this should be implemented by any core feature
-type ChannelHandlerProvider interface {
-	// This is called fast and often -- it must be synchrnozed
+// ChannelNamespaceHandler should be implemented by all core features.
+type ChannelNamespaceHandler interface {
```
I suggest we rename the interface as follows, since the current name doesn't reflect what it does:
```suggestion
// ChannelHandlerFactory should be implemented by all core features.
type ChannelHandlerFactory interface {
```

aknuds1 committed:
Clean up code

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 reviewed (approved):
Looks good AFAICT, but please see suggestion about renaming an interface so it's more understandable.

ryantxu committed:
Update pkg/models/live.go

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Fix build

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 commented:
Thanks for applying my suggestion, fixed up the rest of the code correspondingly :)

ryantxu commented:
nice -- I was about to do it and got conflicts :)

aknuds1 commented:
Fails on e2e tests.

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

torkelo commented:
> In the frontend, we do not want to stop listening immediately when unsubscribed because we need to keep state on refresh and when navigating from page to page 

Can you describe this more? Feels strange to leave things running and pushing data like this and not using RXJS unsubscribe to clean up / remove centrifuge channel subscriptions. 

Does centrifuge keep state? Why do we need another background clean-up process for centrifuge channels?

torkelo commented on public/app/plugins/datasource/grafana/datasource.ts:
```diff
@@ -6,54 +6,63 @@ import {
+      }
+    }
+    // With a single query just return the results
+    if (queries.length === 1) {
+      return queries[0];
+    }
+    if (queries.length > 1) {
+      // HELP!!!
+      return queries[0];
+    }
+    return of(); // nothing
```
not sure I understand this part, and since we are returning getLiveMeasurementsObserver and it calls getStream on the stream handler, when this is unsubscribed by Grafana why is not the centrifuge channel subscription also then removed?

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

ryantxu commented:
> Why do we need another background clean-up process for centrifuge channels?

The primary use case I am hoping to solve is using grafana in the frontend to collect points that are streamed to it.  If we disconnect and reconnect every time an rxjs subscription ends/starts we will lose info.  This happens on dashboard refresh, setting a template variable, navigating pages etc.  We want to be able to connect to a stream and keep using it as we navigate around various dashboards using the same data.  In https://github.com/grafana/grafana/pull/28416, I am exploring how to best solve this --- but it looks like we are not actually unsubscribing everything :(  still need to dig

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

ryantxu committed:
Merge remote-tracking branch 'origin/master' into live-measurments

ryantxu committed:
set the stringField

faxm0dem commented:
I'm confused about the streaming support implementation.
How does this PR relate to https://grafana.com/docs/grafana/latest/developers/plugins/build-a-streaming-data-source-plugin/ ?
